amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -114,12 +126,107 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
+            currentSequenceNumber = ((FanoutRecordsRetrieved)recordsRetrievedContext.getRecordsRetrieved()).continuationSequenceNumber();
+            // Update the triggering flow for post scheduling upstream request.
+            flowToBeReturned = recordsRetrievedContext.getRecordFlow();
+            // Try scheduling the next event in the queue, if available.
+            if (recordsDeliveryQueue.peek() != null) {
+                subscriber.onNext(recordsDeliveryQueue.peek().getRecordsRetrieved());
+            }
+        } else {
+            // Check if the mismatched event belongs to active flow. If publisher receives an ack for a
+            // missing event in active flow, then it means the event was already acked or cleared
+            // from the queue due to a potential bug.
+            if (flow != null && recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier()
+                    .equals(flow.getSubscribeToShardId())) {
+                log.error(
+                        "{}: Received unexpected ack for the active subscription {}. Throwing.  ",
+                        shardId, recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier());
+                throw new IllegalStateException("Unexpected ack for the active subscription");
+            }
+            // Otherwise publisher received a stale ack.
+            else {
+                log.info("{}: Publisher received an ack for stale subscription {}. Ignoring.", shardId,
+                        recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier());
+            }
+        }
+        return flowToBeReturned;
+    }
+
+    // This method is not thread-safe. You need to acquire a lock in the caller in order to execute this.
+    @VisibleForTesting
+    void bufferCurrentEventAndScheduleIfRequired(RecordsRetrieved recordsRetrieved, RecordFlow triggeringFlow) {
+        final RecordsRetrievedContext recordsRetrievedContext =
+                new RecordsRetrievedContext(recordsRetrieved, triggeringFlow, Instant.now());
+        try {
+            // Try enqueueing the RecordsRetrieved batch to the queue, which would throw exception on failure.
+            // Note: This does not block wait to enqueue.
+            recordsDeliveryQueue.add(recordsRetrievedContext);
+            // If the current batch is the only element in the queue, then try scheduling the event delivery.
+            if (recordsDeliveryQueue.size() == 1) {
+                subscriber.onNext(recordsRetrieved);
+            }
+        } catch (IllegalStateException e) {
+            log.warn("{}: Unable to enqueue the payload due to capacity restrictions in delivery queue with remaining capacity {} ",
+                    shardId, recordsDeliveryQueue.remainingCapacity());
+            throw e;
+        } catch (Throwable t) {
+            log.error("{}: Unable to deliver event to the shard consumer.", shardId, t);
```

Comment:
Do you have any use case where record context might be helpful? I feel these two exceptions will be thrown due to capacity constraint, rather than due to the record itself. Let me know if you feel otherwise.